File: Data Structures & Algorithms/word-break/submission-6.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = {len(s) : True}
        def dfs(i):
            if i in memo:
                return memo[i]
            
            for w in wordDict:
                if ((i + len(w)) <= len(s) and 
                     s[i : i + len(w)] == w
                ):
                    if dfs(i + len(w)):
                        memo[i] = True
                        return True
            memo[i] = False
            return False
        
        return dfs(0)

        dp = [False] * (len(s) + 1)
        dp[len(s)] = True

        def recurse(i):
            if dp[i] == True:
                return dp[i]

            for word in wordDict:
                if ((i + len(word)) <= len(s) and s[i : i + len(word)] == word):
                    if recurse(i + len(word)):
                        dp[i] = True
                        return True

            dp[i] = False
            return False

        return recurse(0)

        def recurse(i):
            if i == len(s):
                return True

            for word in wordDict:
                if ((i + len(word)) <= len(s) and s[i : i + len(word)] == word):
                    if recurse(i + len(word)):
                        return True

            return False

        return recurse(0)

File: Data Structures & Algorithms/climbing-stairs/submission-6.cpp
class Solution {
public:
    unordered_map<int, int> dp;

    int dfs(int i, int n) {

        if (i == n) {return 1;}
        if (i > n) {return 0;}

        if (dp.find(i) != dp.end()) {
            return dp[i];
        }

        dp[i] = (dfs(i + 1, n) + dfs(i + 2, n));
        return dp[i];
    }

    int climbStairs(int n) {
        return dfs(0, n);
    }
};
